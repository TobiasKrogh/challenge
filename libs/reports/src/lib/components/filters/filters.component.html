<ng-container [formGroup]="form">
  <select class="form-select" formControlName="projectId">
    <option value="">
      {{ 'REPORTS.ALL_PROJECTS' | translate }}
    </option>
    <option *ngFor="let project of projectsAsArray" [value]="project.projectId">
      {{ project.name }}
    </option>
  </select>

  <select class="form-select" formControlName="gatewayId">
    <option value="">
      {{ 'REPORTS.ALL_GATEWAYS' | translate }}
    </option>
    <option *ngFor="let gateway of gatewaysAsArray" [value]="gateway.gatewayId">
      {{ gateway.name }}
    </option>
  </select>

  <input
    type="hidden"
    formControlName="from"
    ngbDatepicker
    #fromCalendar="ngbDatepicker"
    container="body"
    placement="bottom-right"
    [positionTarget]="fromDateButton"
    [footerTemplate]="fromFooter"
    [minDate]="minDate"
    [maxDate]="maxDate"
    [startDate]="minDate"
  />
  <button
    #fromDateButton
    class="date-picker"
    (click)="fromCalendar.toggle()"
    type="button"
  >
    {{ form.from ? (form.from | formatDate) : 'From Date' }}
  </button>

  <input
    type="hidden"
    formControlName="to"
    ngbDatepicker
    #toCalendar="ngbDatepicker"
    container="body"
    placement="bottom-right"
    [positionTarget]="toDateButton"
    [footerTemplate]="toFooter"
    [minDate]="form.from ?? minDate"
    [maxDate]="maxDate"
  />
  <button
    #toDateButton
    class="date-picker"
    (click)="toCalendar.toggle()"
    type="button"
  >
    {{ form.to ? (form.to | formatDate) : 'To Date' }}
  </button>
</ng-container>

<button type="button" (click)="report()">
  {{ 'REPORTS.GENERATE' | translate }}
</button>

<ng-template #fromFooter>
  <div class="date-reset">
    <button (click)="form.resetControl('from'); fromCalendar.close()">
      {{ 'RESET' | translate }}
    </button>
  </div>
</ng-template>

<ng-template #toFooter>
  <div class="date-reset">
    <button (click)="form.resetControl('to'); toCalendar.close()">
      {{ 'RESET' | translate }}
    </button>
  </div>
</ng-template>
