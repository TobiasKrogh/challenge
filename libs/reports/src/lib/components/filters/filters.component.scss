:host {
  display: flex;
  height: 2rem;
  gap: 1.5rem;
}

select {
  background-color: var(--hex-color-cyan);
  background-image: url(/assets/images/icons/arrow-down.v1.svg);
  border-radius: var(--border-radius--default);
  color: var(--hex-color-white);
  font: var(--font-chart-filter-action);
  padding-top: 0;
  padding-bottom: 0;
  position: relative;
  width: 8rem;

  &:focus {
    box-shadow: none;
  }
}

.date {
  &-picker {
    background-color: var(--hex-color-cyan);
    background-image: url(/assets/images/icons/calendar.v1.svg);
    background-repeat: no-repeat;
    background-position: right 0.75rem center;
    text-align: left;
  }

  &-reset {
    display: flex;
    justify-content: flex-end;
    margin: 0.5rem;
  }
}

button {
  background-color: var(--hex-color-blue--medium);
  border-radius: var(--border-radius--default);
  color: var(--hex-color-white);
  font: var(--font-chart-filter-action);
  height: 2rem;
  padding: 0 0.625rem;
  width: 8rem;
}
