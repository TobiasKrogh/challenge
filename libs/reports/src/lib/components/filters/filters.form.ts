import { FormControl, FormGroup } from '@angular/forms';
import { NgbDateStruct } from '@ng-bootstrap/ng-bootstrap';
import { IFilters } from '@challenge/api';
import { normalizeDate } from '@challenge/shared';

export class FiltersForm extends FormGroup {
  constructor() {
    super({
      gatewayId: new FormControl(''),
      projectId: new FormControl(''),
      from: new FormControl(''),
      to: new FormControl(''),
    });
  }

  public get normalized(): IFilters {
    return Object.fromEntries(
      new Map(
        Object.entries(this.controls)
          .filter(([, { value }]) => !!value)
          .map(([key, control]) => [key, normalizeDate(control.value)])
      )
    );
  }

  public get from(): NgbDateStruct | undefined {
    return this.value.from ?? undefined;
  }

  public get to(): NgbDateStruct | undefined {
    return this.value.to ?? undefined;
  }

  public resetControl(target: string): void {
    return this.patchValue({ [target]: null });
  }
}
