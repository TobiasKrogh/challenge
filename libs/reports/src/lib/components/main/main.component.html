<ng-container
  *ngIf="{
    gateways: gateways$ | async,
    projects: projects$ | async
  } as resolved"
>
  <header>
    <shared-message
      class="message-title"
      [title]="'HOME.TITLE' | translate"
      [text]="'HOME.TEXT' | translate"
    ></shared-message>

    <reports-filters
      (filtersChange)="filters$.next($event); update($event)"
      [gateways]="resolved.gateways"
      [projects]="resolved.projects"
    ></reports-filters>
  </header>

  <div class="content">
    <ng-container *ngIf="report$ | async as report; else noReportGenerated">
      <reports-output
        [filters]="filters$ | async"
        [gateways]="resolved.gateways"
        [projects]="resolved.projects"
        [report]="report"
      ></reports-output>
    </ng-container>

    <ng-template #noReportGenerated>
      <shared-message
        class="message-no-report"
        [title]="'HOME.EMPTY.TITLE' | translate"
        [text]="'HOME.EMPTY.TEXT' | translate"
      >
        <img
          class="message-no-report-image"
          src="/assets/images/no-report.v1.svg"
          alt=""
          width="403"
          height="172"
        />
        <img
          class="message-no-report-icon"
          src="/assets/images/icons/reports.v1.svg"
          alt=""
        />
      </shared-message>
    </ng-template>
  </div>
</ng-container>
