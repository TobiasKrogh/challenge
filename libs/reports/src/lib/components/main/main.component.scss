:host {
  display: flex;
  flex: 1;
  flex-direction: column;
  padding-top: 2rem;
}

header {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
}

.message {
  &-title {
    margin-bottom: 0.5rem;
  }

  &-no-report {
    position: relative;
    text-align: center;

    &-image {
      margin-top: 3.125rem;
    }

    &-icon {
      bottom: 36px;
      position: absolute;
      right: 147px;
    }
  }
}

.content {
  align-items: center;
  display: flex;
  flex: 1;
  justify-content: center;
  padding: 2rem 0;
}
