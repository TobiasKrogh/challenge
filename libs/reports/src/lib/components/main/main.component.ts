import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
import { DataService, IFilters } from '@challenge/api';
import { ReplaySubject } from 'rxjs';

@Component({
  selector: 'reports-main',
  templateUrl: './main.component.html',
  styleUrls: ['./main.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class MainComponent implements OnInit {
  protected readonly filters$ = new ReplaySubject<IFilters>(1);

  protected readonly gateways$ = this.dataService.gateways$;
  protected readonly projects$ = this.dataService.projects$;
  protected readonly report$ = this.dataService.report$;

  constructor(private readonly dataService: DataService) {}

  public ngOnInit(): void {
    this.dataService.loadGateways();
    this.dataService.loadProjects();
  }

  public update(filters: IFilters): void {
    this.dataService.report(filters);
  }
}
