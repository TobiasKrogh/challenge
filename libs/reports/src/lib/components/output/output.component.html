<ng-container
  *ngIf="reportByGateway.length || reportByProject.length; else noData"
>
  <main class="data-display legend">
    <h3>
      <ng-container *ngIf="filters?.projectId as projectId; else allProjects">
        {{ projects?.[projectId]?.name }}
      </ng-container>
      <ng-template #allProjects>
        {{ 'REPORTS.ALL_PROJECTS' | translate }}
      </ng-template>
      |
      <ng-container *ngIf="filters?.gatewayId as gatewayId; else allGateways">
        {{ gateways?.[gatewayId]?.name }}
      </ng-container>
      <ng-template #allGateways>
        {{ 'REPORTS.ALL_GATEWAYS' | translate }}
      </ng-template>
    </h3>

    <section *ngFor="let report of reportByGateway">
      <ng-container
        *ngTemplateOutlet="
          table;
          context: { name: report.gateway.name, report }
        "
      ></ng-container>
    </section>

    <section *ngFor="let report of reportByProject">
      <ng-container
        *ngTemplateOutlet="
          table;
          context: { name: report.project.name, report }
        "
      ></ng-container>
    </section>
  </main>

  <aside *ngIf="showChart; else footer">
    <ng-container *ngIf="reportByGateway.length">
      <ng-container
        *ngTemplateOutlet="
          chartWithFooter;
          context: {
            sources: reportByGateway,
            suffix: 'REPORTS.TOTAL_GATEWAY' | translate
          }
        "
      ></ng-container>
    </ng-container>

    <ng-container *ngIf="reportByProject.length">
      <ng-container
        *ngTemplateOutlet="
          chartWithFooter;
          context: {
            sources: reportByProject,
            suffix: 'REPORTS.TOTAL_PROJECT' | translate
          }
        "
      ></ng-container>
    </ng-container>
  </aside>
</ng-container>

<ng-template #footer>
  <footer class="data-display legend">
    <reports-amount
      [amount]="total"
      [suffix]="'REPORTS.TOTAL' | translate"
    ></reports-amount>
  </footer>
</ng-template>

<ng-template #chartWithFooter let-sources="sources" let-suffix="suffix">
  <reports-chart [sources]="sources" [total]="total"></reports-chart>

  <footer class="data-display legend">
    <reports-amount [amount]="total" [suffix]="suffix"></reports-amount>
  </footer>
</ng-template>

<ng-template #table let-report="report" let-name="name">
  <div
    *ngIf="!isSingleGatewayAndProject"
    class="data-display legend"
    (click)="report.open = report.data.length > 1 ? !report.open : report.open"
  >
    <h4>
      {{ name }}
      <reports-amount
        [amount]="report.total"
        [suffix]="'REPORTS.TOTAL' | translate"
      ></reports-amount>
    </h4>
  </div>
  <table *ngIf="report.open">
    <colgroup>
      <col class="col-25" />
      <col *ngIf="showGateway" class="col-25" />
      <col class="col-25" />
      <col class="col-25" />
    </colgroup>
    <thead>
      <tr>
        <td>Date</td>
        <td *ngIf="showGateway" class="align--centered">Gateway</td>
        <td class="align--centered">Transaction ID</td>
        <td class="align--right">Amount</td>
      </tr>
    </thead>
    <tbody>
      <tr *ngFor="let entry of report.data; trackBy: trackByPaymentId">
        <td>{{ entry.created | formatDate }}</td>
        <td *ngIf="showGateway" class="align--centered">
          {{ gateways?.[entry.gatewayId]?.name }}
        </td>
        <td class="align--centered">{{ entry.paymentId }}</td>
        <td class="align--right">
          <reports-amount [amount]="entry.amount"></reports-amount>
        </td>
      </tr>
    </tbody>
  </table>
</ng-template>

<ng-template #noData>
  <div class="message-no-data">
    <shared-message
      [title]="'REPORTS.NO_DATA.TITLE' | translate"
      [text]="'REPORTS.NO_DATA.TEXT' | translate"
    ></shared-message>
  </div>
</ng-template>
