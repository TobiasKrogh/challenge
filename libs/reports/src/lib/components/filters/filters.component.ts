import {
  ChangeDetectionStrategy,
  Component,
  EventEmitter,
  Input,
  Output,
} from '@angular/core';
import { NgbDateStruct } from '@ng-bootstrap/ng-bootstrap';
import {
  IFilters,
  IGateway,
  IProject,
  TIndexedGateways,
  TIndexedProjects,
} from '@challenge/api';
import { FiltersForm } from './filters.form';

@Component({
  selector: 'reports-filters',
  templateUrl: './filters.component.html',
  styleUrls: ['./filters.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FiltersComponent {
  @Output() filtersChange = new EventEmitter<IFilters>();

  @Input() gateways: TIndexedGateways | null = {};
  @Input() projects: TIndexedProjects | null = {};

  public minDate: NgbDateStruct = {
    year: 2021,
    month: 1,
    day: 1,
  };

  public maxDate: NgbDateStruct = {
    year: 2021,
    month: 12,
    day: 31,
  };

  public form = new FiltersForm();

  public generate(): void {
    this.filtersChange.emit(this.form.normalized);
  }

  public get gatewaysAsArray(): IGateway[] {
    return Object.values(this.gateways ?? {});
  }

  public get projectsAsArray(): IProject[] {
    return Object.values(this.projects ?? {});
  }
}
