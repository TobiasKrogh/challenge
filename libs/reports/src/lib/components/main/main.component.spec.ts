import { NgIf } from '@angular/common';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { TranslatePipe } from '@ngx-translate/core';
import { DataService } from '@challenge/api';
import { MessageComponent } from '@challenge/shared';
import { MockComponents, MockPipes } from 'ng-mocks';
import { ReplaySubject } from 'rxjs';
import {
  MOCK_GATEWAY,
  MOCK_PROJECT,
  MOCK_REPORT,
} from '../../../testing/data.mocks';
import { FiltersComponent } from '../filters/filters.component';
import { OutputComponent } from '../output/output.component';

import { MainComponent } from './main.component';

describe('MainComponent', () => {
  let component: MainComponent;
  let fixture: ComponentFixture<MainComponent>;

  let dataService: DataService;

  const gateways$ = new ReplaySubject(1);
  const projects$ = new ReplaySubject(1);
  const report$ = new ReplaySubject(1);

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [NgIf, MessageComponent],
      providers: [
        {
          provide: DataService,
          useValue: {
            loadGateways: jest.fn(),
            loadProjects: jest.fn(),
            report: jest.fn(),
            gateways$,
            projects$,
            report$,
          },
        },
      ],
      declarations: [
        MockComponents(FiltersComponent, OutputComponent),
        MockPipes(TranslatePipe),
        MainComponent,
      ],
    }).compileComponents();

    dataService = TestBed.inject(DataService);

    fixture = TestBed.createComponent(MainComponent);
    component = fixture.componentInstance;

    gateways$.next([MOCK_GATEWAY]);
    projects$.next([MOCK_PROJECT]);
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  describe('When rendering', () => {
    it('Then it loads gateways and projects', () => {
      fixture.detectChanges();
      expect(dataService.loadGateways).toHaveBeenCalledWith();
      expect(dataService.loadProjects).toHaveBeenCalledWith();
    });

    describe('And no report was generated', () => {
      it('Then it matches snapshot', () => {
        fixture.detectChanges();
        expect(fixture.debugElement).toMatchSnapshot();
      });
    });

    describe('And report was generated', () => {
      it('Then it matches snapshot', () => {
        report$.next([MOCK_REPORT]);
        fixture.detectChanges();
        expect(fixture.debugElement).toMatchSnapshot();
      });
    });
  });

  describe('When updating', () => {
    it('Then it refreshes report', () => {
      component.update({});
      expect(dataService.report).toHaveBeenCalledWith({});
    });
  });
});
