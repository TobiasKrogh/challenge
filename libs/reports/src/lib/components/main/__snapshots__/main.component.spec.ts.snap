// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainComponent When rendering And no report was generated Then it matches snapshot 1`] = `
DebugElement {
  "nativeNode": <div
    id="root2"
  >
    <header>
      <shared-message
        class="message-title title--with-subline"
      >
        
        
      </shared-message>
      <reports-filters />
    </header>
    <div
      class="content"
    >
      <shared-message
        class="message-no-report title--with-subline"
      >
        
        
        <img
          alt=""
          class="message-no-report-image"
          height="172"
          src="/assets/images/no-report.v1.svg"
          width="403"
        />
        <img
          alt=""
          class="message-no-report-icon"
          src="/assets/images/icons/reports.v1.svg"
        />
      </shared-message>
      
      
    </div>
    
    
  </div>,
}
`;

exports[`MainComponent When rendering And report was generated Then it matches snapshot 1`] = `
DebugElement {
  "nativeNode": <div
    id="root3"
  >
    <header>
      <shared-message
        class="message-title title--with-subline"
      >
        
        
      </shared-message>
      <reports-filters />
    </header>
    <div
      class="content"
    >
      <reports-output />
      
      
      
    </div>
    
    
  </div>,
}
`;
